packages/eslint-plugin/src/rules/no-non-null-asserted-nullish-coalescing.ts

```diff
@@ -0,0 +1,94 @@
+import { TSESTree, TSESLint } from '@typescript-eslint/experimental-utils';
+import {
+  Definition,
+  DefinitionType,
+  Variable,
+} from '@typescript-eslint/scope-manager';
+import * as util from '../util';
+
+function hasAssignmentBeforeNode(
+  variable: Variable,
+  node: TSESTree.Node,
+): boolean {
+  return (
+    variable.references.some(
+      ref => ref.isWrite() && ref.identifier.range[1] < node.range[1],
+    ) ||
+    variable.defs.some(
+      def =>
+        isDefinitionWithAssignment(def) && def.node.range[1] < node.range[1],
+    )
+  );
+}
+
+function isDefinitionWithAssignment(definition: Definition): boolean {
+  if (definition.type !== DefinitionType.Variable) {
+    return false;
+  }
+
+  const variableDeclarator = definition.node;
+  return variableDeclarator.definite ?? variableDeclarator.init !== null;
+}
+
+export default util.createRule({
+  name: 'no-non-null-asserted-nullish-coalescing',
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Disallows using a non-null assertion in the left operand of the nullish coalescing operator',
+      category: 'Possible Errors',
+      recommended: false,
+      suggestion: true,
+    },
+    messages: {
+      noNonNullAssertedNullishCoalescing:
+        'The nullish coalescing operator is designed to handle undefined and null - using a non-null assertion is not needed.',
+      suggestRemovingNonNull: 'You should remove the non-null assertion.',
```

Comment:
nit - this is the text that shows up for the suggestion fix in the IDE.
So the wording should say what the fix is doing instead of suggesting what the user might do.

```suggestion
      suggestRemovingNonNull: 'Remove the non-null assertion.',
```